Jenkinsfile

```diff
@@ -0,0 +1,307 @@
+      when {
+        // If none of these are set, then we can't do the curl to get the list of
+        // github labels. Currently we then fall through, warn we have not done the
+        // check, and then run the CI. We *could* skip the CI in this situation if
+        // we wanted to...
+        allOf {
+          expression { GITHUB_API_TOKEN != null }
+          expression { env.ghprbGhRepository != null }
+          expression { env.ghprbPullId != null }
+        }
+      }
+      steps {
+        script {
+          DID_SKIPTEST='true'
+          // Curl the github labels for this PR (the ghprb plugin does not supply them directly),
+          // and then check if we have the 'fast track the CI' label set.
+          json=sh(script: "/usr/bin/curl -H \"Authorization: Basic ${GITHUB_API_TOKEN}\" https://api.github.com/repos/${env.ghprbGhRepository}/issues/${env.ghprbPullId}/labels", returnStdout: true).trim()
+
+          // Convert the labels JSON to a list of maps, one map per label
+          labs = readJSON text: "${json}"
+          // Check if we have the 'skip the CI' label applied
+          labs.each { labmap ->
+            if ( labmap['name'] == SKIPLABEL ) {
+              echo "Found CI skip label, skipping further tests"
+              SKIPTESTS='true'
+            }
+          }
+          if ( SKIPTESTS != 'true' ) {
+            echo "CI fastpath not set, running full pipeline"
+          }
+        }
+      }
+    }
+
+    // Warn when we failed to do the precheck. This is not idea, as it will appear
+    // as an extra stage. it would be great if the above 'when' check had an 'else'
+    // clause we could use, but afaict, it does not.
+    stage('CheckPrecheck') {
+      agent { label "master" }
+      when {
+        expression { DID_SKIPTEST == 'false' }
+      }
+      steps {
+        echo "Warning, skipped CI skip check"
+      }
+    }
+
+    // We will always need some of these to run the static checks.
+    // We need to git clone the:
+    // - tests repo so we can use some of its CI scripts
+    // - rutime repo to get the versions.yaml file
+    // - ci repo to get the Jenkins test matrix YAML
+    stage('Setup test repo environment') {
+      agent { label "master" }
+      steps {
+        sh '''
+          git clone "https://${tests_repo}.git" "${WORKSPACE}/${tests_repo_dir}"
+          git clone "https://${runtime_repo}.git" "${WORKSPACE}/${runtime_repo_dir}"
+          git clone "https://${ci_repo}.git" "${WORKSPACE}/${ci_repo_dir}"
+        '''
+      }
+    }
+
+    // Always run the static checks, even if we have a 'skip-ci' label.
+    // Later we could add a 'really-skip-ci' label if we needed etc. to
+    // not even run the static checks??
+    // If the PR does not pass the static checks,
+    // then the rest of the CI checks will be skipped by Jenkins.
+    stage('Static checks') {
+      agent { label "master" }
+      steps {
+        script {
+          echo "In static check"
```

Comment:
Wot no `.ci/static-checks.sh`? :smile:

Reply:
Sigh - look, this is an **example** of how we might do this right now. I'm looking for high level architectural feedback....

Reply:
Easy :tiger: ! It was a joke!